src/Chapter2.hs

```diff
@@ -620,7 +625,26 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven' :: [a] -> [a]
```

Comment:
I've come up with two very similar solutions:
1.  match on list structure
2. match on list length

Reply:
Very creative ideas! 🎨 🧠 

From these two options that you choose, I see that both of then are using `reverse`. It is slow and expensive as you noticed.

In option 1 you walk the list and you need to walk it again at the end, which is very time-consuming.

In option 2, you use `length` at each step, but length walks the whole list in order to get the list length, so your function would take quadratic time + reverse - one additional path through the list at the end.

Also, both of these solutions missing an opportunity for streaming - this is very important for lazy languages. This means that you won't be able to get half-results on the fly, so you won't be able to do something like:

```haskell
take 5 (takeEven (cycle [1])
```

With your solution, this function never ends execution, but with streaming solution, it could return you `[1,1,1,1,1]` straightaway.

Note that you can implement this function with the simple pattern matching and this would solve the mentioned above problems 🙂 It requires fewer function calls and will be more efficient both by time and by memory. 🕐 
As you need to take every second element, it is possible to represent with a few pattern matching clauses 🙂